### PR TITLE
[macsec]: Recover neighbor devices by sanity check

### PR DESCRIPTION
**Pull request by Pterosaur**
Signed-off-by: Ze Gan <ganze718@gmail.com>

### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
The neighbor devices may fail to initialize MACsec due to an abnormal condition that the MACsec interfaces weren't cleanupped in the last test.

#### How did you do it?
If "MACsec" is enabled to check whether the MACsec interface isn't in neighbor device. If not, recover the neighbor devices by config reload.

#### How did you verify/test it?
Check Azp

### CODE REVIEW

**Comment by wangxin** [comment]
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The detected issues may be old or new. For new issues, please try to fix them.

For old issues, it is not mandatory to fix them because they were not caused by this change. It is unfair to blame
author of this pull request. But if you can take extra effort to fix the old issues as well, that would be great!

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/plugins/sanity_check/checks.py:30:34: E231 missing whitespace after ','<br/>tests/common/plugins/sanity_check/checks.py:874:1: E302 expected 2 blank lines, found 1<br/>tests/macsec/__init__.py:19:1: F401 'tests.common.plugins.sanity_check.sanity_check' imported but unused<br/>tests/macsec/__init__.py:23:1: E302 expected 2 blank lines, found 1<br/>tests/macsec/__init__.py:87:41: E127 continuation line over-indented for visual indent<br/>tests/macsec/__init__.py:88:41: E127 continuation line over-indented for visual indent<br/>tests/macsec/__init__.py:89:41: E127 continuation line over-indented for visual indent<br/>tests/macsec/__init__.py:139:121: E501 line too long (149 > 120 characters)<br/>tests/macsec/__init__.py:139:148: E203 whitespace before ':'<br/>tests/macsec/__init__.py:154:121: E501 line too long (144 > 120 characters)<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

**Comment by wangxin** [comment]
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The detected issues may be old or new. For new issues, please try to fix them.

For old issues, it is not mandatory to fix them because they were not caused by this change. It is unfair to blame
author of this pull request. But if you can take extra effort to fix the old issues as well, that would be great!

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/macsec/__init__.py:20:1: F401 'tests.common.plugins.sanity_check.sanity_check' imported but unused<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

File: tests/common/plugins/sanity_check/recover.py [hunk 1]
```diff
@@ -103,13 +105,27 @@ def _neighbor_vm_recover_bgpd(node=None, results=None):
...
+
+
+def neighbor_vm_restore(duthost, nbrhosts, tbinfo, result):
     logger.info("Restoring neighbor VMs for {}".format(duthost))
     mg_facts = duthost.get_extended_minigraph_facts(tbinfo)
     vm_neighbors = mg_facts['minigraph_neighbors']
     if vm_neighbors:
-        results = parallel_run(_neighbor_vm_recover_bgpd, (), {}, nbrhosts.values(), timeout=300)
-        logger.debug('Results of restoring neighbor VMs: {}'.format(json.dumps(dict(results))))
+        if result["check_item"] == "bgp":
```
Comment (yejianquan) [inline tests/common/plugins/sanity_check/recover.py:119]: Would it be better to not affect original logic?
by:

if result["check_item"] == "neighbor_macsec_empty":
            unhealthy_nbrs = []
            for name, host in nbrhosts.items():
                if name in result["unhealthy_nbrs"]:
                    unhealthy_nbrs.append(host)
            parallel_run(_neighbor_vm_recover_config, (), {}, unhealthy_nbrs, timeout=300)
            logger.debug('Results of restoring neighbor VMs: {}'.format(unhealthy_nbrs))
else: 
            results = parallel_run(_neighbor_vm_recover_bgpd, (), {}, nbrhosts.values(), timeout=300)
            logger.debug('Results of restoring neighbor VMs: {}'.format(json.dumps(dict(results))))

File: tests/macsec/__init__.py [hunk 2]
```diff
@@ -46,27 +49,27 @@ def pytest_generate_tests(self, metafunc):
             metafunc.parametrize('macsec_profile',
                                  [self.macsec_profiles[x] for x in profiles],
                                  ids=profiles,
-                                 scope="session")
+                                 scope="module")
 
-    @pytest.fixture(scope="session")
+    @pytest.fixture(scope="module")
```
Comment (yejianquan) [inline tests/macsec/__init__.py:54]: Just curious, does "session" or 'module" make any difference in our pytest framework?
We always run one module in a session, by "-c module" parameter.

Comment (Pterosaur) [reply tests/macsec/__init__.py:54]: No difference. But the fixture, sanity_check, is module level. Meanwhile, the session fixtures are always executed before the module fixtures. So, I change macsec fixture to guarantee the MACsec setup will execute after sanity_check.

**Review by yejianquan — APPROVED** [review]
LGTM

**Comment by yejianquan** [comment]
@Pterosaur commit 0288924 can't be cherry-picked into 202012 or 202205 branches due to conflicts.
Please fix the conflict if you think 202012 and 202205 need it, 
if not, please remove the "Request for 202012/202205 branch" tags.

**Pull request merged**